Add function to checkout the jobs repository

ColinKinloch opened the pull request:
This branch adds `crate::clone_git_repository` to clone git repos generally and `Job::clone_git_repository` to clone specifically defined by the gitlab job.

`clone_git_repository` is heavily based on [`PrepareFetch::fetch_only`](https://github.com/GitoxideLabs/gitoxide/blob/b10cc6f0f2c3757677d9fdb3c01196d371ac6631/gix/src/clone/fetch/mod.rs#L66-L212)

Since the gitoxide [`PrepareFetch`](https://docs.rs/gix/latest/gix/clone/struct.PrepareFetch.html) and [`PrepareCheckout`](https://docs.rs/gix/latest/gix/clone/struct.PrepareCheckout.html) structs delete the target path when they are dropped I opted to checkout the repository to a temporary directory, then move it's contents to the path provided to the target directory.

This adds a rather unwieldy `GitCheckoutError` enum. gitoxides `Error` types triggers the `result_large_err` clippy warning.

This requires rust version `1.73.0` or later.

Fixes https://github.com/collabora/gitlab-runner-rs/issues/21

ColinKinloch committed:
Add the ability to checkout git repositories

* `clone_git_repository` to clone arbitrary repo
* `Job::clone_git_repository` to clone the jobs repo

andrewshadura reviewed (approved):
This looks okay, but I haven’t gone deeply into details. I trust you tested this code well? (It would be great if it also included some automated tests.)

sjoerdsimons commented on gitlab-runner/src/lib.rs:
```diff
@@ -454,3 +457,148 @@ impl Runner {
 }
+
+// TODO: Should clone_git_repository be async
+//   gitoxide doesn't currently have async implemented for http backend
```
fwiw i'd prefer to stay with gitoxide to avoid more C dependences in the core of gitlab-runner. But yes it should be done in a dedicated thread if there is no async api

ColinKinloch committed:
typo: "Failed to create build dir"

ColinKinloch committed:
Support git clone by commit id SHA

Also throw an error if the provided HEAD ref cannot be parsed.

ColinKinloch committed:
Offload gitoxide to a tokio blocking thread

ColinKinloch commented on gitlab-runner/src/lib.rs:
```diff
@@ -454,3 +457,148 @@ impl Runner {
+    }
+
+    if let Some(depth) = depth {
+        if depth > 0 {
```
Nice

ColinKinloch commented on gitlab-runner/src/lib.rs:
```diff
@@ -454,3 +457,148 @@ impl Runner {
+                refspec::parse::Operation::Fetch,
+            ) {
+                refspecs.push(refspec.to_owned());
+            }
```
I've updated it to throw an error.
It doesn't really make sense to checkout if the user specified a head that we've failed to parse.

refi64 committed:
Make GitCheckoutError's inner errors boxed

Otherwise, clippy gets quite upset because the error type is so large
(>200 bytes).

refi64 committed:
Clarify comment on git credentials helper

refi64 committed:
Fix incorrect interpretation of allow_git_fetch

This is set if the runner isn't allowed to reuse an existing repository
by re-running a fetch there, which is unrelated to the ability to clone
itself.

refi64 committed:
Add comment clarifying meaning of the depth field

It's a bit confusing that this automatically considers the corresponding
CI variable, but some other settings (such as the git strategy allowing
repo reuse) do not.

refi64 committed:
mock: Add the ability to customize more GitInfo fields

refi64 committed:
Add tests for clone_git_repository

refi64 committed:
Make git checkout APIs private

No real reason for them to be public, and it frees us from handling a
lot of edge cases and allows for some significant (future) code cleanup.

refi64 committed:
Relocate git clone code to job.rs

If this is private, and only used by Job, it makes sense for it to be in
the same place.

refi64 committed:
Make clone_git_repository natively cancel-safe

Given that the example didn't even use the cancellation token properly,
it's better for cancellation to be handled internally.

refi64 committed:
Reduce scope of git cloning code to only handle GitLab's inputs

The `sha` field is never going to contain a non-SHA, among other things,
so this code can be wittled down quite a bit now.

refi64 committed:
Use more gitoxide helpers for the cloning process

Much of this can be replaced with helpers from the crate itself.

refi64 committed:
Add support for avoiding clones when GIT_STRATEGY=none/empty

In these cases, we just return None to signal back to the caller that
the clone couldn't be performed for a non-error reason.
